user-console/src/main/java/org/pentaho/mantle/client/ui/tabs/MantleTabPanel.java

```diff
@@ -94,6 +175,45 @@ public void addTab( String text, String tooltip, boolean closeable, Widget conte
     if ( getSelectedTab() == null ) {
       selectTab( tab );
     }
+    MantleTabMenuItem menuItem = createTabMenuItem( tab );
+    menuItemHashMap.put( tab, menuItem );
+    tabsSubMenuBar.addItem( menuItem );
+    contextMenuRefreshThreshold( true );
+    updateTabMenuText( tab );
+  }
+
+  public void renameMenuTab( PentahoTab tab ){
+    menuItemHashMap.get( tab ).setText( tab.getLabelText() );
+    updateTabMenuText( tab );
+  }
+
+  public void updateTabMenuText( PentahoTab tab ) {
+    if ( tab == null ) {
+      tabsMenuBar.addStyleName( EMPTY_TABS_MENU );
+      tabsMenuItem.setText( "" );
+    } else {
+      tabsMenuBar.removeStyleName( EMPTY_TABS_MENU );
+      String tabMenuText;
+      int tabCount = getTabCount();
+      if ( tabCount > 1 ) {
+        int tabIndex = tabsSubMenuBar.getItemIndex( menuItemHashMap.get( tab ) ) + 1;
+        tabMenuText = "Tab (" + tabIndex + "/" + tabCount + ") " + tab.getLabelText();
+      } else {
+        tabMenuText = tab.getLabelText();
+      }
+      tabsMenuItem.setText( tabMenuText );
+    }
+  }
+
+  private MantleTabMenuItem createTabMenuItem( MantleTab tab ) {
+    MantleTabMenuItem tabMenuItem = new MantleTabMenuItem( tab );
+    return tabMenuItem;
+  }
+
+  private void contextMenuRefreshThreshold( boolean added ) {
+    if ( ( added && menuItemHashMap.size() == 2 ) || ( !added && menuItemHashMap.size() == 1 ) ) {
```

Comment:
magic numbers `2` and `1`, add a javadoc comment as to why 2 and 1 are chosen for MantleTabPane#contextMenuRefreshThreshold.